### PR TITLE
Enumerable methods

### PR DESCRIPTION
**Pull request by alvarorf**
In this project we have done the following:
- Read the documentation to understand the core functionality of each, each_with_index, select, all?, any?, none?, count, map and inject ruby built-in enumerable methods.
- Implement our own enumerable methods to mimic the functionality of the above-mentioned methods
- Test our enumerable methods against the built-in methods

### CODE REVIEW

File: enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+
+    return to_enum unless block_given?
+
+    i = 0
+    while i <= length - 1
+      yield(self[i])
+      i += 1
+    end
+    self
+  end
```
Comment (bertil291utn) [inline enumerables.rb]: ## Calling a given block
-  When passing a block as a parameter should the process start working

Example:

```ruby
new_proc = proc { |num| num < (0 + 9) / 2 }
p Range.new(5, 50).each(&new_proc)
# => expcted 5..50
p Range.new(5, 50).my_each(&new_proc)
# got => nil
```


## Hashes, range, and array self
- [x] Test with hashes and ranges, then compare with original `each` Ruby method
- [x] Fix your method to work with the tests. Follow this [sample](https://gist.github.com/bertil291utn/5a7e6a73351c38ceb9781a9a37b58cf6)

**Example:**


```ruby
(0..9).my_each{
# or
Range.new(0, 9)

# HASHES
{ a: 1, b: 2, c: 3, d: 4, e: 5 }
```

File: enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    block = lambda { |obj| obj } unless block_given?
+
+    i = 0
+    while i < length
+      return true if block.call(self[i])
+
+      i += 1
+    end
+    false
+  end
```
Comment (bertil291utn) [inline enumerables.rb]: ## Class integer
- Passing a class integer as parameter.
 - [x] Please fix this issue 
Example:
```ruby
p %w[cat mouse].any?(Integer)
# => expected: false
p %w[cat mouse].my_any?(Integer)
# => got: true
```

File: enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+
+  def my_none?(*arg, &block)
+    !my_any?(*arg, &block)
+  end
```
Comment (bertil291utn) [inline enumerables.rb:148]: ## With regular expressions
- Check also with regular expression

Example:

```ruby
p %w[cat mouse].none?(/z/)
# => expected: true
p %w[cat mouse].my_none?(/z/)
# => expected: false
```

## With class
```ruby
p %w[cat mouse].none?(5)
# => expected: true
p %w[cat mouse].my_none?(5)
# => expected: false
```

File: enumerables.rb [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+    end
+    block = lambda { |obj| obj } unless block_given?
+
+    i = 0
+    while i < length
+      count += 1 if block.call(self[i])
+      i += 1
+    end
+    count
+  end
```
Comment (bertil291utn) [inline enumerables.rb]: ## With range new class

- Please check it works also with Range new classes

```ruby
p Range.new(5, 50).count
# => expected: 46
p Range.new(5, 50).my_count
# => got: undefined local variable or method `length' for 5..50:Range (NameError)
```

## How many items passing an integer

```ruby
p [1, 3, 78, 90, 3].count(3)
# => expected: 2
p [1, 3, 78, 90, 3].my_count(3)
# => got: 7
```

File: enumerables.rb [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+          while i < length
+            result = yield(result, self[i])
+            i += 1
+          end
+          result
+        end
+      end
+    end
+  end
+end
```
Comment (bertil291utn) [inline enumerables.rb]: ## Return Local Jump Error
- When there's no block given should return Local Jump Error

```ruby
p [1, 3, 78, 90, 3].inject
# => expected: no block given (LocalJumpError)
p [1, 3, 78, 90, 3].my_inject
# => got: nil
```

## Specified a symbol
-  When a symbol is specified without an initail value combines each element of the collection by applying the symbol
```ruby
p [1, 3, 78, 90, 3].inject(:+)
# => expected: 175
p [1, 3, 78, 90, 3].my_inject(:+)
# => got: nil
```


## Symbol with an initial value with a range new class
- When a symbol is specified with an initial value combines each element of the collection by applying the symbol

```ruby
p Range.new(1, 10).inject(2, :*)
# => expected: 7257600
p Range.new(1, 10).my_inject(2, :*)
# => got: nil
```

**Review by bertil291utn — CHANGES_REQUESTED** [review]
# **Status:** Changes requested 🟠 
<p align="center"> image <br><strong>Not yet</strong> </p>

# Hey guys, you've been doing a great job, and also I hope you enjoyed coding. 
### And @alvarorf  it's nice to see you again man 👍 🙌 

### Your work is amazing you master it GitHub Flow you have a professional README file. 

However, in order to accomplish with all requirements, you required to do these changes, listed below:

- [x] Edit pull request message following this [guideline](https://cdn.filestackcontent.com/osvFJEkQHWQsgDe0PJoQ)


### OPTIONAL
- [x] README.md file add bullets to enumerate each method
Example:
```markdown
Welcome to our project "Advanced Building Blocks:Enumerable Methods". 

The purpose of this project was to develop our own implementation of some of the Enumerable methods such as each, 
- each_with_index
- map 
- inject
- none?
- count
-  all?
- any?
-  select.
```

### Summary
- Please check with `Range.new(1,40)` and regular expression 
-  Don't forget to work with inline comments below this comment box.

***

Guys work in the requested changes above and request another review.
Good luck. 

> “Never stop, never conform until what is good becomes even better, and what’s better becomes excellent.”

File: enumerables.rb [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    i = 0
+    while i < length
+      yield(self[i], i)
+      i += 1
+    end
+    self
+  end
```
Comment (elmejdki) [inline enumerables.rb]: ## #my_each_with_index:

- [x] allow #my_each_with_index to work with ranges: `undefined local variable or method `length' for 5..50:Range`

File: enumerables.rb [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+    block = lambda { |obj| obj } unless block_given?
+
+    i = 0
+    while i < length
+      return false unless block.call(self[i])
+
+      i += 1
+    end
+    true
+  end
```
Comment (elmejdki) [inline enumerables.rb]: ## #my_all?:

```ruby

HIGHEST_VALUE = 9
range = Range.new(5, 50)
false_block = proc { |num| num > HIGHEST_VALUE }
words = %w[dog door rod blade]
array = [2, 34, 43, 54, 68, 93, 3, 14, 62, 51, 28, 38, 34]

```

- [x] returns false if the block returns false or nil::range

```ruby

  range.my_all?(&false_block) === range.all?(&false_block) => false

```

- [x] when a Regex is passed as an argument returns true if all of the collection matches the Regex

```ruby

  words.my_all?(/d/) === words.all?(/d/) => false

```

- [x] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

```ruby

  array.my_all?(3) === array.all?(3) => false

```

File: enumerables.rb [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+    block = lambda { |obj| obj } unless block_given?
+
+    i = 0
+    while i < length
+      return true if block.call(self[i])
+
+      i += 1
+    end
+    false
+  end
```
Comment (elmejdki) [inline enumerables.rb]: ## #my_any?:

```ruby

HIGHEST_VALUE = 9
range = Range.new(5, 50)
false_block = proc { |num| num > HIGHEST_VALUE }
words = %w[dog door rod blade]
array = [2, 34, 43, 54, 68, 93, 3, 14, 62, 51, 28, 38, 34]

```

- [x] returns false if the block ever returns a value false or nil::range

```ruby 

  range.my_any?(&false_block) === range.any?(&false_block) => false

```

- [x] when a Regex is passed as an argument returns true if any of the collection matches the Regex

```ruby 
  words.my_any?(/d/) === words.any?(/d/) => false

```

- [x] when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern

```ruby 
  words.my_any?('cat') === words.any?('cat') => false

```

File: enumerables.rb (hunk 3 above)
Comment (elmejdki) [reply enumerables.rb:148]: work on this also as mentioned by the previous TSE

File: enumerables.rb [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+      to_a.my_each { |elt| res.push(yield(elt)) }
+      res
+    end
+    return unless is_a? Array
+
+    return to_enum unless block_given?
+
+    my_each { |elt| res.push(yield(elt)) }
+    res
+  end
```
Comment (elmejdki) [inline enumerables.rb]: ## #my_map:

```ruby
  HIGHEST_VALUE = 9 
  LOWEST_VALUE = 0
  block = proc { |num| num < (LOWEST_VALUE + HIGHEST_VALUE) / 2 }
  my_proc = proc { |num| num > 10 }
```

- [x] returns a new array with the results of running a given block once for every element in enum::range

```ruby

range.my_map(&block) === range.map(&block) => false
```

- [x] executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.

```ruby

  array.my_map(my_proc) { |num| num < 10 } === array.map(&my_proc) => false
```

File: enumerables.rb [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+
+  def my_inject(*arg, &block)
+    return raise LocalJumpError.new "No block given" if (arg.length == 0) && (!block_given?)
+
+    if arg.length > 2
+      return ArgumentError.new "wrong number of arguments (given #{arg.length}, expected 0..2)"
+    end
```
Comment (elmejdki) [inline enumerables.rb]: ## #my_inject:

```ruby
  range = Range.new(5, 50)
```

- [x] when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method:range

```ruby

  range.my_inject(:*) === range.inject(:*) => false

# also

  range.my_inject(2, :*) === range.inject(2, :*) => false
```

**Review by elmejdki — CHANGES_REQUESTED** [review]
# status: require changes :recycle: 
## Great job so far @team :clap: :+1: :100:, but we still need some more work :hammer: to get this project approved :checkered_flag:

- [x] check the comments below for more info about the requirements

### Please fix all the required changes before you request another code review.
## Happy Coding :tophat:

**Review by kubilaycaglayan — CHANGES_REQUESTED** [review]
## Hello Team 👋
## Status: Changes Required ♻️

Great job so far 👍🎯
I will point out the exact points that you should improve, and give examples... 😊
Let's do it together ⏬

### my_each_with_index
- [x]  when a block is given when self is a hash behaves like #each_with_index calls the given block once for each element in self

💡
```ruby
{1 => 'Kubilay', 2 => 'Pink Floyd'}.each_with_index do |pair|
  p pair
end

{1 => 'Kubilay', 2 => 'Pink Floyd'}.my_each_with_index do |pair|
  p pair
end

# OUTPUT

# [1, "Kubilay"]
# [2, "Pink Floyd"]
# nil
# "Kubilay"
```

### my_all?
- [x]  when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

💡
```ruby
array = []
3.times { array << 1991 }

p array.all?(1991)
p array.my_all?(1991)

# OUTPUT
# true
# false
```

### my_any?

- [x]  when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern

💡 Similar 👆

### my_none?
- [x]  when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern

💡 Similar 👆

### That's it! After making the changes you can request another code review 😊
With all my respect 🙏

**Review by euqueme — APPROVED** [review]
> ### Pull Request reviewed by Maria Eugenia Queme P. (TSE)

### Hello @Blaise-Shyaka and @alvarorf :lollipop: :balloon: 

### Thank you for doing all the changes requested by the previous reviewers :1st_place_medal: :1st_place_medal: all your enumerables are working incredibly :sparkles: so please be proud of your great work you may merge this branch and move on Congratulations!! :tada: :tada: :tada: 

### Keep up the good work :clinking_glasses: 

PROJECT STATUS: APPROVED :heavy_check_mark:

**Pull request merged**